xarray/core/combine.py

```diff
@@ -323,16 +324,20 @@ def _dataarray_concat(arrays, dim, data_vars, coords, compat,
         raise ValueError('data_vars is not a valid argument when '
                          'concatenating DataArray objects')
 
-    datasets = []
-    for n, arr in enumerate(arrays):
-        if n == 0:
-            name = arr.name
-        elif name != arr.name:
-            if compat == 'identical':
-                raise ValueError('array names not identical')
-            else:
-                arr = arr.rename(name)
-        datasets.append(arr._to_temp_dataset())
+    name = result_name(arrays)
+    names = [arr.name for arr in arrays]
+    if compat == 'identical' and len(set(names)) != 1:
+        raise ValueError(
+            "compat='identical', but array names {!r} are not identical"
+            .format(names if len(names) <= 10 else sorted(set(names)))
+        )
+    datasets = [arr.rename(name)._to_temp_dataset() for arr in arrays]
+
+    if isinstance(dim, str) and len(set(names) - {None}) == len(names) \
```

Comment:
can we not do `None in set(names)`?

Reply:
Not for this - we're checking that names is a list of unique strings.